atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/query/GetQuerySpec.java

```diff
@@ -47,40 +48,46 @@
             + "ORDER BY column1, column2 ASC "
             + "LIMIT 1;";
 
-    @org.immutables.value.Value.Auxiliary
-    public abstract ByteBuffer row();
-
-    @org.immutables.value.Value.Auxiliary
-    public abstract ByteBuffer column();
+    @Override
+    public String formatQueryString() {
+        return String.format(
+                QUERY_FORMAT,
+                cqlQueryContext().keySpace(),
+                AbstractKeyValueService.internalTableName(cqlQueryContext().tableReference()));
+    }
 
-    @org.immutables.value.Value.Auxiliary
-    public abstract long humanReadableTimestamp();
+    @Override
+    public Statement makeExecutableStatement(PreparedStatement preparedStatement) {
+        return preparedStatement.bind()
+                .setBytes("row", getQueryParameters().row())
+                .setBytes("column", getQueryParameters().column())
+                .setLong("timestamp", getQueryParameters().queryTimestamp());
+    }
 
     @Override
     public ConsistencyLevel queryConsistency() {
         return ConsistencyLevel.LOCAL_QUORUM;
     }
 
     @Override
-    public final Supplier<RowStreamAccumulator<Optional<Value>>> rowStreamAccumulatorFactory() {
+    public Supplier<RowStreamAccumulator<Optional<Value>>> rowStreamAccumulatorFactory() {
         return GetQueryAccumulator::new;
     }
 
-    @Override
-    public String formatQueryString() {
-        return String.format(QUERY_FORMAT, keySpace(), AbstractKeyValueService.internalTableName(tableReference()));
-    }
+    @org.immutables.value.Value.Auxiliary
+    public abstract GetQueryParameters getQueryParameters();
 
-    @Override
-    public Statement makeExecutableStatement(PreparedStatement preparedStatement) {
-        return preparedStatement.bind()
-                .setBytes("row", row())
-                .setBytes("column", column())
-                .setLong("timestamp", queryTimestamp());
-    }
+    @org.immutables.value.Value.Immutable
+    public interface GetQueryParameters {
+        ByteBuffer row();
+
+        ByteBuffer column();
 
-    private long queryTimestamp() {
-        return ~humanReadableTimestamp();
+        long humanReadableTimestamp();
+
+        default long queryTimestamp() {
+            return ~humanReadableTimestamp();
+        }
     }
 
     @ThreadSafe
```

Comment:
In #accumulateRowStream, you are verifying that the resultValue is not set, and then mapping anyway. You should just assign it. Better still, I would have written the query specs for single row things first, and if we ever decide to move more queries, we can refactor and introduce that capability.

We could then remove all the paging and page iteration code there, or keep it but specialize it for a single page.

Reply:
The `RowStreamAccumulator` does not force you to process pages, multiple rows etc. If you know that you'll get all the data in the first call of `accumulateRowStream` you can do whatever you want in the impl. That is why I intended `GetQueryAccumulator` to be a private internal class of GetQuerySpec, it is an implementation detail based on the query we wrote so we can make it as specific to the query as we want. We can discuss in person if this is not a right way to organize code.

I agree we can specialize the code more without paging etc. I think it makes more sense to do it after I switch to the new driver version since it has a built in support for async.